### PR TITLE
Add download from Azure Blobs to DM.

### PR DESCRIPTION
**Pull request by mamokarz**

### CODE REVIEW

File: STMicroelectronics/STM32L4_L4+/app/azure_config.h [hunk 1]
```diff
@@ -23,7 +23,7 @@ typedef enum
 // Azure IoT Dynamic Provisioning Service
 //    Define this to use the DPS service, otherwise direct IoT Hub
 // ----------------------------------------------------------------------------
-#define ENABLE_DPS
+//#define ENABLE_DPS
```
Comment (mamokarz) [inline STMicroelectronics/STM32L4_L4+/app/azure_config.h:26]: Undo this change.

File: STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/inc/stm32l475_flash_driver.h [hunk 2]
```diff
@@ -0,0 +1,8 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license.
+// See LICENSE file in the project root for full license information.
+
```
Comment (mamokarz) [inline STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/inc/stm32l475_flash_driver.h:4]: Protect against double inclusion. #Resolved

File: STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/inc/stm32l475_flash_driver.h [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+#include "stm32l4xx_hal.h"
+
```
Comment (mamokarz) [inline STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/inc/stm32l475_flash_driver.h:9]: Add external C for C++ #Resolved

File: DCF_Demo.md [hunk 4]
```diff
@@ -422,200 +250,113 @@ az iot hub invoke-device-method -n [name-of-iothub] -d [name-of-device] --mn "ip
...
 
+Install you package to a remote device running DCF using invoke-device-method:
 
-## Test the New Package
 <details>
 <summary>Click here for Portal/bash commands</summary>
 <br>
 
-Query for existing interfaces on the device 
+Query for existing interfaces on the device, if you have sprinkler.1 installed, uninstall it before try to install the new package from Blobs.
```
Comment (smcd253) [inline DCF_Demo.md]: May want to include link to uninstall instructions here. #Resolved

File: DCF_Demo.md [hunk 5]
```diff
@@ -674,7 +414,7 @@ The commands are different for Powershell because you have to use ` to escape an
 
 <br>
 
-Query for existing interfaces on the device 
+Query for existing interfaces on the device, if you have sprinkler.1 installed, uninstall it before try to install the new package from Blobs.
```
Comment (smcd253) [inline DCF_Demo.md]: nit: before trying to install #Resolved

File: DCF_Demo.md [hunk 6]
```diff
@@ -9,282 +9,185 @@ You can use the Azure Portal or an instance on Powershell/bash/other local envir
...
-### Built in example
-The current example contains 3 packages already built in. Cypher, in 2 versions, and Sprinkler, in one version. The following commands will drive the installation, utilization and des-installation of those packages. 
+### In Memory example
+
+Install an **In Memory** package assume that the binary is already in the MCU memory when the install command arrives. The main goal of this installation is to debug package's code. 
+If properly configured, all debug tools applicable to a code built in the OS can be equally applied to a package using the DCF in memory installation.
+
+The current example produces the key_vault_v1i1 and sprinkler_v1i1 packages. The CMake will create the key_vault_v1i1.elf and key_vault_v1i1.bin for key_vault_v1i1 package in ".\build\packages\key_vault_v1i1" directory, and sprinkler_v1i1.elf and sprinkler_v1i1.bin for sprinkler_v1i1 package in ".\build\packages\sprinkler_v1i1" directory. 
+
+For now, both packages was created to run in a specific memory address, key_vault_v1i1 shall be installed in the address 0x08050000, and sprinkler_v1i1 in the address 0x08057000.
```
Comment (smcd253) [inline DCF_Demo.md]: nit: For now, both packages were created... #Resolved

File: STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/az_ulib_dm.c [hunk 7]
```diff
@@ -84,75 +82,54 @@ AZ_NODISCARD az_result az_ulib_dm_deinit(void)
   return _az_ulib_dm_interface_unpublish();
 }
 
-static az_result install_built_in(az_span package_name)
+static az_result install_in_memory(void* base_address, az_span package_name)
 {
   _az_PRECONDITION_NOT_NULL(_az_dm_cb);
```
Comment (mamokarz) [inline STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/az_ulib_dm.c:89]: Move this test to the API. #Resolved

File: STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/stm32l475_flash_driver.c [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+        {
+            HAL_FLASH_Lock();
+            return status;
+        }
+        remainder_count = -1;
+    }
+
+    for (; size > 8; size -= 8, destination_ptr += 8, source_ptr += 8)
+    {
+        status = HAL_FLASH_Program(FLASH_TYPEPROGRAM_DOUBLEWORD, (uint32_t)(destination_ptr), (uint64_t)*(uint32_t*)source_ptr | ((uint64_t)*(uint32_t*)(source_ptr + 4)) << 32);
```
Comment (mamokarz) [inline STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/stm32l475_flash_driver.c]: Break this line. #Resolved

**Review by cindydeng1998 — APPROVED** [review]
🌞

**Review by smcd253 — COMMENTED** [review]
documentation nits

File: DCF_Demo.md (hunk 4 above)
Comment (mamokarz) [reply DCF_Demo.md]: Sounds good, thanks

**Review by smcd253 — APPROVED** [review]
lgtm
